File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_19.py
# **********
# ***    ***
# **      **
# *        *
# *        *
# **      **
# ***    ***
# **********

# BRUTE FORCE
n = 5
# Upper half
for i in range(n , 1 , -1):
    stars = "*" * i
    spaces = " " * (2 * (n - i))
    print(f"{stars}{spaces}{stars}")
# Lower half
for i in range(1 , n+1 ):
    stars = "*" * i
    spaces = " " * (2 * (n - i))
    print(f"{stars}{spaces}{stars}")
    
# BETTER APPROACH
n = 5
rows = []

# upper half
for i in range(n, 1, -1):
    stars = "*" * i
    spaces = " " * (2 * (n - i))
    rows.append(f"{stars}{spaces}{stars}")

# lower half
for i in range(1, n + 1):
    stars = "*" * i
    spaces = " " * (2 * (n - i))
    rows.append(f"{stars}{spaces}{stars}")
    
# Output all rows
for r in rows:
    print(r)
    
# BEST APPROACH
n = 5

for k in range(2 * n - 1):
    # Map 0..(n-2) → n..2 and (n-1)..(2n-2) → 1..n
    if k < n - 1:
        i = n - k
    else:
        i = k - (n - 2)

    stars = "*" * i
    spaces = " " * (2 * (n - i))
    print(f"{stars}{spaces}{stars}")

File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_19.cpp

// **********
// ***    ***
// **      **
// *        *
// *        *
// **      **
// ***    ***
// **********

// BRUTE FORCE
#include <iostream>
#include <string>
#include <vector>
using namespace std;
int main()
{
    int n = 5;
    // Upper half
    for (int i = n; i > 1; i--)
    {
        string stars(i, '*');
        string spaces(2 * (n - i), ' ');
        cout << stars << spaces << stars << endl;
    }
    // Lower half
    for (int i = 1; i <= n; i++)
    {
        string stars(i, '*');
        string spaces(2 * (n - i), ' ');
        cout << stars << spaces << stars << endl;
    }

    cout << endl;
    return 0;
}

// BETTER APPROACH
#include <iostream>
#include <string>
#include <vector>
using namespace std;
int main()
{
    int n = 5;
    vector<string> rows;

    // upper half
    for (int i = n; i > 1; i--)
    {
        string stars(i, '*');
        string spaces(2 * (n - i), ' ');
        rows.push_back(stars + spaces + stars);
    }

    // lower half
    for (int i = 1; i <= n; i++)
    {
        string stars(i, '*');
        string spaces(2 * (n - i), ' ');
        rows.push_back(stars + spaces + stars);
    }

    // Output all rows
    for (string r : rows)
        cout << r << endl;

    cout << endl;
    return 0;
}

// BEST APPROACH
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;

    for (int k = 0; k < 2 * n - 1; k++)
    {
        int i;

        // Map upper half and lower half
        if (k < n - 1)
            i = n - k;
        else
            i = k - (n - 2);

        string stars(i, '*');
        string spaces(2 * (n - i), ' ');
        cout << stars << spaces << stars << endl;
    }

    return 0;
}
